### PR TITLE
Fix HPA sample sanitization

### PR DESCRIPTION
**Pull request by jbartosik**
**What this PR does / why we need it**: @mwielgus pointed out a case when HPA fails as a result of my changes to HPA algorithm:
- Have pods that use a lot of CPU during initilization, become ready right after they initialize,
- Trigger a scale up,
- When new pods become ready will will count their usage (even though it's not related to any work that needs doing),
- This triggers another scale up, even though existing pods can handle work, no problem.

The fix is:
- Use all samples for non-cpu metrics.
- Only use CPU samples if:
  - Pod is ready and was started more than 2 minutes ago, or
  - Pod is unready and last readiness change happened more than 10s after it was started.

Reasoning behind this in: https://docs.google.com/document/d/1UdtYedhmCxjaJIQi6hwJMY0eHQQKxlVD8lSHZC1BPOA/edit

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

**Release note**:
```release-note
Replace scale up forbidden window with disregarding CPU samples collected when pod was initializing.
```

### CODE REVIEW

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
you need to add what the fix is to the PR description.  This definitely needs a release note.  All algorithm changes need release notes.

Separately, hard coding a startup window is not acceptable.  Some pods start up within 10s, others take 5 minutes.  We can't just arbitrarily assume 2 minutes is correct.

**Comment by jbartosik** [comment]
Done. I Updated description. And also I changed implementation a little (to match the conclusions in the doc).

**Comment by jbartosik** [comment]
@DirectXMan12 please take a look

**Comment by jbartosik** [comment]
@DirectXMan12 I added a commit turning hard coded durations into flags.

I rebased because the new commit had a merge conflict.

File: pkg/controller/podautoscaler/replica_calculator.go
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/replica_calculator.go]: a) you need a comment as to why this is two minutes
b) if this is based on the metrics-server collection window, it shouldn't be constant -- it should be using the reported window size from metrics-server.  If this is "just a guess", that's not a good answer.

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]

File: pkg/controller/podautoscaler/replica_calculator.go
Comment (jbartosik) [reply pkg/controller/podautoscaler/replica_calculator.go]: thought you left this comment in a different commit and were referring to
values HPA uses (which were hard coded in the previous commit).

Those are just values I use for testing, it doesn't matter much what exactly they are.

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (DirectXMan12) [inline pkg/apis/componentconfig/v1alpha1/types.go]: is this actually based on the metric window?  if so, it should use the metric window reported by the custom metrics API, not be hardcoded.

Comment (jbartosik) [reply pkg/apis/componentconfig/v1alpha1/types.go]: I agree in principle. But that would be significantly more complicated (currently client doesn't return metric window so I would have to change client, add more tests, and modify existing tests).

I need to merge this in 2 days so I'd rather not do it in this PR. Can we settle for a TODO. I'd send a PR addressing the TODO when I'm back from vacation (and merge it after the freeze).

File: pkg/controller/podautoscaler/replica_calculator.go
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/replica_calculator.go]: extraneous log statement.

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
found an extraneous log line.  modulo that and the changes around window calculation we discussed, this looks good.

File: cmd/kube-controller-manager/app/options/options_test.go
Comment (mwielgus) [inline cmd/kube-controller-manager/app/options/options_test.go]: 5s?

File: cmd/kube-controller-manager/app/options/options_test.go
Comment (mwielgus) [inline cmd/kube-controller-manager/app/options/options_test.go]: 5s?

File: pkg/apis/componentconfig/v1alpha1/defaults.go
Comment (mwielgus) [inline pkg/apis/componentconfig/v1alpha1/defaults.go]: 10 sec only?

Comment (jbartosik) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: Yes. This is what we assume is max duration between pod starting and the first time it get readiness assigned

Comment (jbartosik) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: It was <2s for the pods I looked at

File: pkg/apis/componentconfig/v1alpha1/defaults.go
Comment (DirectXMan12) [inline pkg/apis/componentconfig/v1alpha1/defaults.go]: if this is based on HPA processing interval + window, and we are going to actually collect window from the API, why do we need this separate option?  Why not just use the HPA processing window, so we don't have this option at all?  (that's what I was trying to suggest before)

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]

File: pkg/apis/componentconfig/v1alpha1/defaults.go
Comment (jbartosik) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: We want to avoid using samples with window overlapping with pod initialization. If we use sample after {end of initialization time} + {sample window} we will achieve that.

{sample window} is hardcoded for now, I'll send PR for using actual value tomorrow.

We cannot know {end of initialization time} without storing per-pod information which is not feasible. So I'm going with a heuristic - initialization takes no more than a HorizontalPodAutoscalerCPUTaintPeriod (1 minute by default).

**Comment by mwielgus** [comment]
@DirectXMan12 
I took a liberty of LGTMing the PR to get it through the door, based on @jbartosik claims of your verbal conditional LGTM. As @jbartosik is going for vacation next week @krzysztof-jastrzebski will be doing all of the follow-up PRs to adjust readiness-related logic to both your and our preferences.

File: pkg/apis/componentconfig/v1alpha1/defaults.go
Comment (mwielgus) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: Let's make it 30second, just in case Kubelet is stalled.

**Comment by jbartosik** [comment]
@wojtek-t Done.

**Comment by wojtek-t** [comment]
/lgtm

[Though those validation exceptions need api-reviewer approval].

**Comment by wojtek-t** [comment]
/hold

@thockin - for final approval of the exceptions.

**Comment by thockin** [comment]
I am not sure what I am asked to approve?  what is triggering the violation?

**Comment by jbartosik** [comment]
@thockin  I'm adding new flags for HPA. They require corresponding fields in componentconfig v1alpha1 API. All such fields have an exception (I think because they start with a capital letter).

**Comment by jbartosik** [comment]
@thockin - please remove hold on this PR @wojtek-t placed it to make sure you approve it.

**Comment by jbartosik** [comment]
@DirectXMan12 @mwielgus please merge this PR

**Comment by DirectXMan12** [comment]
> As @jbartosik is going for vacation next week @krzysztof-jastrzebski will be doing all of the follow-up PRs to adjust readiness-related logic to both your and our preferences.

As long as the PRs are posted before code slush that's acceptable.

**Pull request merged**
